M2/Macaulay2/packages/Macaulay2Doc/functions/document-doc.m2

```diff
@@ -0,0 +1,640 @@
+    ExampleFiles=>List
+      see @TOH ExampleFiles@
+  Consequences
+    Item
+      formatted documentation is created and stored in the @TO2 {"currentPackage", "current package"}@
+  Description
+    Text
+      There are two basic types of documentation:
+
+      @OL {
+          {"Documenting a function or a method, as in ", TO (resolution, Module), "."},
+          {"documenting an overview or functions with methods, as in ", TO "chain complexes",
+              " or ", TO resolution, " respectively."}}@
+
+      The headings @TO Usage@, @TO BaseFunction@, @TO Inputs@, @TO Outputs@, and @TO Consequences@,
+      are useful only for documentation of the first type, as well as in a call to @TO SYNOPSIS@.
+
+    Tree
+      :Templates for using the @TO (document, List)@ method
+        "package documentation template"
+        "function documentation template"
+        "optional argument documentation template"
+        "overview documentation template"
+
+    Text
+      Here is a basic template:
+    Code
+      EXAMPLE { PRE ////
+          document {
+              Key          => {myFunc, (myFunc, MyType), [myFunc, MyOption], ...},
+              Headline     => "one line description of myFunc", -- not needed for overviews
+              BaseFunction => "myFunc", -- usually not needed
+              Usage        => "myFunc x",
+              Inputs       => { inputs, ... },
+              Outputs      => { outputs, ... },
+              SourceCode   => { METHOD1, ... }, -- usually not needed
+              Consequences => { side-effects, ... },
+              "A list of strings containings names of files in the auxiliary
+              directory of the package can go here; the files will be visible
+              in the current directory while example code is run.",
+              ExampleFiles => {"datafile1", "datafile2"},
+              "There can be explanatory prose here in the form of a hypertext list.",
+              EXAMPLE lines //////
+                m2code
+                m2code
+                m2code
+              //////,
+              "There can be explanatory prose here in the form of a hypertext list.",
+              Caveat       => {"warning"},
+              SeeAlso      => {"other things"},
+              Subnodes     => {
+                  "subheading a",
+                  TO "node 1",
+                  TO "node 2",
+                  "subheading b",
+                  TO "node 3",
+                  ...
+                  },
+              }////}
+    Tree
+      :Other Functions generating @TO "hypertext list format"@ that you may use:
+        SYNOPSIS
+        EXAMPLE
+  SeeAlso
+    "writing documentation"
+    hypertext
+  Subnodes
+    :Documentation templates based on the hypertext list format
+    "package documentation template"
+    "function documentation template"
+    "optional argument documentation template"
+    "overview documentation template"
+    :Special documentation keywords
```

Comment:
In combination with the previous issue, maybe we should not list the optional inputs at all, especially because some of them are not "optional" inputs at all, like Key. What do you think?

Reply:
No, look how much more informative the current way of doing it is:
<img width="784" alt="Screen Shot 2020-09-08 at 1 36 17 PM" src="https://user-images.githubusercontent.com/700228/92509522-fccc6780-f1cf-11ea-94cb-4d8b20e1fd39.png">

Reply:
It's kind of a mess though. I think there's a point after which making a "list" is actually counterproductive. The same information can be provided somewhere else. See the documentation of SimpleDoc keywords as an alternative.

Reply:
> It's kind of a mess though. I think there's a point after which making a "list" is actually counterproductive. The same information can be provided somewhere else. See the documentation of SimpleDoc keywords as an alternative.

That's more work for the author, and is prone to error, if, for example, one is omitted or falsely included.

Reply:
I don't understand why you brought up `installPackage`, when we're discussing `document`. The purpose of optional arguments is completely different in the two, and like I said, in the case of document they're not all optional.

We can spend hours implementing checks or fixing inconsistencies by hand, but at the end of the day if the documentation is to messy to read, it will be ignored.

Reply:
Regardless, we can discuss whether they are documented as optional arguments elsewhere, but for now I don't want to remove the listed keywords at the end, which you consider redundant, because those are added by the `SubNodes` keyword and removing them would ruin the table of contents:
![image](https://user-images.githubusercontent.com/3147718/92580930-a35a4c00-f254-11ea-9a51-4d3412789dd4.png)

Reply:
> I don't understand why you brought up `installPackage`, when we're discussing `document`. The purpose of optional arguments is completely different in the two, and like I said, in the case of document they're not all optional.
> 
> We can spend hours implementing checks or fixing inconsistencies by hand, but at the end of the day if the documentation is to messy to read, it will be ignored.

The pull request makes the optional argument section a lot messier.  Here is a comparison:
<img width="1315" alt="Screen Shot 2020-09-09 at 11 07 21 AM" src="https://user-images.githubusercontent.com/700228/92617061-8b022580-f284-11ea-9788-0ef56ba2c270.png">

Reply:
> Regardless, we can discuss whether they are documented as optional arguments elsewhere, but for now I don't want to remove the listed keywords at the end, which you consider redundant, because those are added by the `SubNodes` keyword and removing them would ruin the table of contents:

Ah, good point.  And for `document`, these aren't really optional arguments, since they are elements of a list rather than a sequence.  So removing this would be fine:
```m2
    Key=>List
      see @TOH Key@
    Headline=>List
      see @TOH Headline@
    BaseFunction=>List
      see @TOH BaseFunction@
    Usage=>List
      see @TOH Usage@
    Inputs=>List
      see @TOH Inputs@
    Outputs=>List
      see @TOH Outputs@
    Consequences=>List
      see @TOH Consequences@
    Acknowledgement=>List
      see @TOH Acknowledgement@
    Contributors=>List
      see @TOH Contributors@
    References=>List
      see @TOH References@
    Caveat=>List
      see @TOH Caveat@
    SeeAlso=>List
      see @TOH SeeAlso@
    Subnodes=>List
      see @TOH Subnodes@
    SourceCode=>List
      a list of functions whose source code should be displayed in the documentation
    ExampleFiles=>List
      see @TOH ExampleFiles@
```

Reply:
> The pull request makes the optional argument section a lot messier. Here is a comparison:

You consider this messy?

![image](https://user-images.githubusercontent.com/3147718/92627247-b559e000-f290-11ea-8e59-bf5fb665b783.png)

> Ah, good point. And for document, these aren't really optional arguments, since they are elements of a list rather than a sequence. So removing this would be fine:

The last two, SourceCode and ExampleFiles, aren't documented separately. Can you document them? I don't get the need for ExampleFiles.

Reply:
So, here's the situation:
1. I hate this: (this is the result of removing the section you quoted above)
![image](https://user-images.githubusercontent.com/3147718/92645244-2b1e7580-f2aa-11ea-973a-15d2511e5856.png)
2. The way to remove it is to have `document` not take any options; this is easy enough:
```m2
document = method(Dispatch => Thing)
document Sequence := args -> document toList args
document List     := args -> ...
```
3. However, if we do that, then nodes such as `[document, Key]` can't exist, which is not unimaginable, but I find having a bunch of loose keywords unpleasant.

I'm okay with not making any other changes and marking this as resolved, perhaps circling back later.

Reply:
Actually, wait, removing the links from the options was stupid. I added that back and modified the remainder of the lines for this node. Marking this as resolved because I think it looks fine now. Feel free to reopen if you disagree.